pkg/shell/hosts.jsx

```diff
@@ -0,0 +1,267 @@
+import cockpit from "cockpit";
+
+import React from 'react';
+import ReactDOM from "react-dom";
+import PropTypes from 'prop-types';
+import { PageSidebar, Button } from '@patternfly/react-core';
+import { EditIcon, MinusIcon } from '@patternfly/react-icons';
+
+import { superuser } from "superuser.jsx";
+import { CockpitNav, CockpitNavItem } from "./nav.jsx";
+
+import { machines } from "machines";
+import { new_machine_dialog_manager } from "machine-dialogs";
+
+import "../../node_modules/@patternfly/patternfly/components/Select/select.scss";
+
+import $ from "jquery"; // Ech... Fixme: We now use the same dialogs as dashboard and we need jQuery to open them
+
+const _ = cockpit.gettext;
+const hosts_sel = document.getElementById("nav-hosts");
+
+class HostsSelector extends React.Component {
+    constructor() {
+        super();
+        this.el = document.createElement("div");
+        this.el.className = "view-hosts";
+    }
+
+    componentDidMount() {
+        hosts_sel.appendChild(this.el);
+    }
+
+    componentWillUnmount() {
+        hosts_sel.removeChild(this.el);
+    }
+
+    render() {
+        const { children } = this.props;
+        return ReactDOM.createPortal(children, this.el);
+    }
+}
+
+function HostLine({ host, user }) {
+    return (
+        <>
+            <span id="current-username" className="username">{user}</span>
+            {user && <span className="at">@</span>}
+            <span className="hostname">{host}</span>
+        </>
+    );
+}
+
+export class CockpitHosts extends React.Component {
+    constructor(props) {
+        super(props);
+
+        this.state = {
+            opened: false,
+            editing: false,
+            current_user: "",
+            current_key: props.machine.key,
+            privileged: false,
+        };
+
+        this.toggleMenu = this.toggleMenu.bind(this);
+        this.filterHosts = this.filterHosts.bind(this);
+        this.onAddNewHost = this.onAddNewHost.bind(this);
+        this.onEditHosts = this.onEditHosts.bind(this);
+        this.onHostEdit = this.onHostEdit.bind(this);
+        this.onRemove = this.onRemove.bind(this);
+
+        this.machines = machines.instance();
+        this.mdialogs = new_machine_dialog_manager(this.machines);
+    }
+
+    componentDidMount() {
+        superuser.addEventListener("changed", () => this.setState({ privileged: superuser.allowed }));
+
+        this.setState({ privileged: superuser.allowed });
```

Comment:
This isn't necessary, and possibly even buggy. `superuser.allowed` starts out as `null`, and `privileged` looks like it should be a boolean. It is correct to start with `false`, and only set it to true once `superuser` initialized, that's what we do on other pages too. I. e. please just drop that line.

Reply:
It is needed, I tried it without and when I initialized it to false, it never called the `"changed"` callback (because when it was imported it was already in the expected state) and thus it never was changed to privileged. I will set it to `!!superuser.allows` to get true/false out of it correctly.

Reply:
Oh, I see my error -- it's because it's in DidMount() instead of the constructor. Then it's necessary, of course.